File: Python/0424LongestRepeatingCharacterReplacement.py
def characterReplacement(s : str, k : int) -> int:
    n = len(s)
    left = 0
    maxFreq = 0
    ans = 0
    freq = [0] * 26

    for right in range(n):
        idx = ord(s[right]) - ord('A')
        freq[idx] += 1
        maxFreq = max(maxFreq, freq[idx])

        while (right - left + 1 - maxFreq) > k:
            idxL = ord(s[left]) - ord('A')
            freq[idxL] -= 1
            left += 1
        
        ans = max(ans, (right - left + 1))
    
    return ans

s = "ABAB"
k = 2
print(characterReplacement(s, k))

File: C/0424LongestRepeatingCharacterReplacement.c
int characterReplacement(char* s, int k) {
    int n = strlen(s);
    int left = 0;
    int maxFreq = 0;
    int ans = 0;

    int freq[26] = {0};

    for (int right = 0; right < n; right++) {
        int idx = s[right] - 'A';
        freq[idx]++;

        if (freq[idx] > maxFreq) {
            maxFreq = freq[idx];
        }

        while (((right - left + 1) - maxFreq) > k) {
            int idxL = s[left] - 'A';
            freq[idxL]--;
            left++;
        }

        int windowLen = right - left + 1;
        if (windowLen > ans) {
            ans = windowLen;
        }
    }

    return ans;
}